ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/metastore/events/CreateFunctionEvent.java

```diff
@@ -0,0 +1,96 @@
+import org.apache.hadoop.hive.metastore.events.PreCreateFunctionEvent;
+import org.apache.hadoop.hive.metastore.events.PreEventContext;
+import org.apache.hadoop.hive.ql.security.authorization.plugin.HiveOperationType;
+import org.apache.hadoop.hive.ql.security.authorization.plugin.HivePrivilegeObject;
+import org.apache.hadoop.hive.ql.security.authorization.plugin.metastore.HiveMetaStoreAuthorizableEvent;
+import org.apache.hadoop.hive.ql.security.authorization.plugin.metastore.HiveMetaStoreAuthzInfo;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/*
+ Authorizable Event for HiveMetaStore operation CreateFunction
+ */
+public class CreateFunctionEvent extends HiveMetaStoreAuthorizableEvent {
+    private static final Logger LOG = LoggerFactory.getLogger(CreateFunctionEvent.class);
+
+    private String COMMAND_STR = "create function";
+
+    public CreateFunctionEvent(PreEventContext preEventContext) {
+        super(preEventContext);
+    }
+
+    @Override
+    public HiveMetaStoreAuthzInfo getAuthzContext() {
+        HiveMetaStoreAuthzInfo ret = new HiveMetaStoreAuthzInfo(preEventContext, HiveOperationType.CREATEFUNCTION, getInputHObjs(), getOutputHObjs(), COMMAND_STR);
+
+        return ret;
+    }
+
+    private List<HivePrivilegeObject> getInputHObjs() { return Collections.emptyList(); }
+
+    private List<HivePrivilegeObject> getOutputHObjs() {
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("==> CreateFunctionEvent.getOutputHObjs()");
+        }
+
+        List<HivePrivilegeObject> ret   = new ArrayList<>();
+        PreCreateFunctionEvent event = (PreCreateFunctionEvent) preEventContext;
+        Function function = event.getFunction();
+        List<ResourceUri> uris   = function.getResourceUris();
+        ret.add(new HivePrivilegeObject(HivePrivilegeObject.HivePrivilegeObjectType.DATABASE, function.getDbName(), null, null, null,
+                HivePrivilegeObject.HivePrivObjectActionType.OTHER, null, null,
+                function.getOwnerName(), function.getOwnerType()));
+        ret.add(new HivePrivilegeObject(HivePrivilegeObject.HivePrivilegeObjectType.FUNCTION, function.getDbName(), function.getFunctionName(), null,
+                null, HivePrivilegeObject.HivePrivObjectActionType.OTHER, null, function.getClassName(), function.getOwnerName(), function.getOwnerType()));
+
+        if (uris != null && !uris.isEmpty()) {
+            for(ResourceUri uri: uris) {
+                ret.add(new HivePrivilegeObject(HivePrivilegeObject.HivePrivilegeObjectType.DFS_URI, null, uri.getUri()));
+            }
+        }
+
+        COMMAND_STR = buildCommandString(COMMAND_STR, function);
+
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("<== CreateFunctionEvent.getOutputHObjs(): ret=" + ret);
+        }
+
+        return ret;
+    }
+
+    private String buildCommandString(String cmdStr, Function function) {
```

Comment:
why are we passing in cmdStr at all. It seems like a constant that can just reference the COMMAND_STR directly?

Reply:
Ack